FreeTubeApp/FreeTube
# In app download

---
In app download
---

**Pull Request Type**
Please select what type of pull request this is:

**Related issue**
related to #1772

**Description**
The first step to add a downloads tab. This feature changes the old mechanism to download media to one where the download is carry in app. From the settings menu, the user can define a working folder or demand to be asking every time where to download the media. The progress is displayed at intervals of around 20% into a toast.

**Screenshots**
![2022-01-11_16-57](https://user-images.githubusercontent.com/19601176/149027908-ae92ae19-cc60-4cc5-991c-7f9ff9d71667.png)


![2021-12-29_08-38](https://user-images.githubusercontent.com/19601176/147668512-89480bdb-6da2-47e3-8939-3fcc534eccc7.png)

![2021-12-29_08-38_1](https://user-images.githubusercontent.com/19601176/147668503-eb19913d-a72a-409a-9cdd-87c347f3a593.png)

![2021-12-29_08-37](https://user-images.githubusercontent.com/19601176/147668529-fbe38bb1-08e0-4406-bb3c-e7b294f4a448.png)





**Testing**
ask for download path

- Go into option
- Click into the download setting
- set to ask for download path
- Click on the download button in a video page 
- Check if a save dialog appears
- Check if some toast displays the progress
- Check if the file is downloaded and functional

select  download path

- Go into option
- Click into the download setting
- Set to do not ask for download path
- click choose the path
- check if a directory dialog appears
- check if the path is written into the display
- Click on the download button in a video page 
- Check if some toast displays the progress
- Check if the file is downloaded and functional

try with multiple downloads links

**Desktop**
 - OS: Ubuntu
- OS Version: 20.04.3 LTS
- FreeTube version: v0.15.1 Beta

**Additional context**
The next step will be to create a download tab, the one after that is to enable the possibility to read the download file instead of streaming the media.
